File: review/DynamicProgramming/2293-coin1.py
[N, K] = input().split(' ')
N = int(N)
K = int(K)
coins = []
for i in range(int(N)):
    coins.append(int(input()))

dp = [0 for i in range(K+1)]
for coin in coins:
    for i in range(K-coin+1):
        if i==0:
            dp[i+coin]+=1
        else:
            dp[i+coin] += dp[i]

print(dp[K])
# // 00:10:33 메모리 초과 -> python으로 교체
# // 00:17:52 틀렸습니다(3%) -> 불필요 print 제거
# 00:20:12 맞았습니다

File: review/DynamicProgramming/2293-coin1.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : require('path').resolve(__dirname, '../../testcase.txt'))
    .toString()
    .trim()
    .split('\n')

const [N,K] = input[0].split(' ').map(Number)
const coins = input.slice(1).map(Number)

const dp = Array(K+1).fill(0)
for(const coin of coins){
    for(let i=coin;i<=K;i++){
        if(i===coin){
            dp[i] += 1
        }else{
            dp[i] += dp[i-coin] ?? 0
        }
    }
}
console.log(dp[K])

// 00:10:33 메모리 초과 -> 
// 